File: Leetcode_240.py
#Time Complexity: O(m+n)
#Space Complexity: O(1)


class Solution:
    def searchMatrix(self, matrix, target: int) -> bool:
        if not matrix or not matrix[0]:
            return False
        rows, cols=len(matrix), len(matrix[0])
        r,c = 0, cols-1
        
        while r<rows and c>=0:
            if matrix[r][c]==target:
                return True
            if target>matrix[r][c]:
                r+=1
            else:
                c-=1
        return False

File: LeetCode240.java

// Time Complexity : O(m+n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int m = matrix.length;
        int n = matrix[0].length;
         int row=0, col =n-1;

         while(row<m && col>=0)  // worst case row can go to m and n can go to 0 (which is  cols)
         {
             if (matrix[row][col]> target)
                 col--;
             else if(matrix[row][col]< target)
             {
                 row++;
             }
            else{
                return true;
            }

         }
   return false;
    }
}